File: TaipeiParkC/TaipeiParkC/ViewController.m
//
//  ViewController.m
//  TaipeiParkC
//
//  Created by Richard on 2018/01/31.
//  Copyright © 2018年 Snowowl. All rights reserved.
//

#import "ViewController.h"
#import "ParkListTableViewCell.h"
#import "ParkDetailViewController.h"

#define QUERY_REMOTE_URL                @"http://data.taipei/opendata/datalist/apiAccess?scope=resourceAquire&rid=bf073841-c734-49bf-a97f-3757a6013812"
#define QUERY_TIMER_INTERVAL            3.0f
#define QUERY_TIMEOUT_INTERVAL          15.0f


@interface ViewController () <NSURLSessionDelegate, UITableViewDataSource, UITableViewDelegate> {
    
    NSTimer *queryTimer;
    NSTimer *timeoutTimer;
    NSTimeInterval  timeout;
    NSMutableData *responseData;

    NSMutableArray *sectionArray;
    NSMutableArray *tableviewArray;
    NSMutableArray *nearbyviewArray;
    
    NSIndexPath *selectedIndexPath;
}

@property (weak, nonatomic) IBOutlet UITableView *aTableView;

@property (weak, nonatomic) IBOutlet UIActivityIndicatorView *activityView;

@property (weak, nonatomic) NSIndexPath *selectedIndexPath;

@end

@implementation ViewController

@synthesize selectedIndexPath = _selectedIndexPath;

- (void)setATableView:(UITableView *)aTableView
{
    _aTableView = aTableView;
    self.aTableView.dataSource = self;
    self.aTableView.delegate = self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    
    self.aTableView.rowHeight = UITableViewAutomaticDimension;
    self.aTableView.estimatedRowHeight = 150;
    
    [self.activityView startAnimating];
    
    [self startQueryTimer];
}



#pragma mark - HTTP Connection

- (void)startConnection {
    NSString *urlString = QUERY_REMOTE_URL;
    
    NSURL *url = [NSURL URLWithString:urlString];

    NSURLRequest *request = [[NSURLRequest alloc] initWithURL:url cachePolicy:NSURLRequestUseProtocolCachePolicy timeoutInterval:QUERY_TIMEOUT_INTERVAL];
    
    NSURLSessionConfiguration *sessionConfiguration = [NSURLSessionConfiguration defaultSessionConfiguration];
    NSURLSession *session = [NSURLSession sessionWithConfiguration:sessionConfiguration
                                                          delegate:self
                                                     delegateQueue:[NSOperationQueue mainQueue]];
    NSURLSessionDataTask *task = [session dataTaskWithRequest:request];
    [task resume];

}

- (void)startQueryTimer {
    if (queryTimer == nil) {
        queryTimer = [NSTimer scheduledTimerWithTimeInterval:QUERY_TIMER_INTERVAL target:self selector:@selector(timerHandler:) userInfo:nil repeats:NO];
    }
}

- (void)timerHandler:(NSTimer*)timer {
    if (timer == queryTimer) {
        [self startConnection];
    } else if (timer == timeoutTimer) {
        
    }
}


#pragma mark - NSURLSessionDelegate

- (void)URLSession:(NSURLSession *)session dataTask:(NSURLSessionDataTask *)dataTask
didReceiveResponse:(NSURLResponse *)response
 completionHandler:(void (^)(NSURLSessionResponseDisposition disposition))completionHandler {
    
    NSHTTPURLResponse *httpResponse = (NSHTTPURLResponse*)response;
    int statusCode = (int)[httpResponse statusCode];
    
    if( statusCode != 200 ) {
        NSLog(@"didReceiveResponse error: %d", statusCode);
        NSLog(@"didReceiveResponse response: %@", response);
    } else if (statusCode == 200) {
        NSLog(@"didReceiveResponse successful");
        responseData = [[NSMutableData alloc] init];
        
        completionHandler(NSURLSessionResponseAllow);
    }

}

- (void)URLSession:(NSURLSession *)session dataTask:(NSURLSessionDataTask *)dataTask didReceiveData:(NSData *)data {
    [responseData appendData:data];
}

- (void)URLSession:(NSURLSession *)session task:(NSURLSessionTask *)task didCompleteWithError:(NSError *)error {
    if (error) {
        NSLog(@"didCompleteWithError: %@", error);
    } else {
        NSString *jsonString = [[NSString alloc] initWithData:responseData encoding:NSUTF8StringEncoding];
        
        //NSLog(@"didCompleteWithError data: %@", jsonString);
        
        [self parseNprocess:jsonString];

    }
}


#pragma mark - Parsing JSON Data

- (void)parseNprocess:(NSString*)jsonString {
    NSData *jsonData = [jsonString dataUsingEncoding:NSUTF8StringEncoding];
    id json = [NSJSONSerialization JSONObjectWithData:jsonData options:0 error:nil];
    
    //NSLog(@"JSON Dictionary: %@", json);
    
    sectionArray = [[NSMutableArray alloc] init];
    tableviewArray = [[NSMutableArray alloc] init];
    
    NSDictionary *result = [json objectForKey:@"result"];
    
    NSArray *results = [result objectForKey:@"results"];
    
    __block int nImageCount = 0;
    __block int nThumbCount = 0;
    
    if (results.lastObject) {
        NSDictionary *currentItem = [results objectAtIndex:0];
        [sectionArray addObject:currentItem];
        
        NSMutableArray *nearbyArray = [[NSMutableArray alloc] init];
        
        __block int nLoadedImageCount = 0;
        
        for (NSDictionary *item in results) {
            NSString *imageURL = [[item objectForKey:@"Image"] description];
            NSString *parkName = [[item objectForKey:@"ParkName"] description];
            NSString *name = [[item objectForKey:@"Name"] description];
            NSString *introduction = [[item objectForKey:@"Introduction"] description];
            NSString *openTime = [[item objectForKey:@"OpenTime"] description];
            NSString *yearBuilt = [[item objectForKey:@"YearBuilt"] description];
            
            NSLog(@"\nImage URL: %@,\nPark Name:%@,\nName: %@,\nOpen Time: %@,\nYear Build: %@,\nIntroduction: %@\n",
                  imageURL, parkName, name, openTime, yearBuilt, introduction);
            
            if (![imageURL isEqualToString:@""]) {
                nImageCount++;
            }
            
            __block NSMutableDictionary *cellItem = [NSMutableDictionary dictionaryWithDictionary:item];
            
            dispatch_queue_t queue = dispatch_queue_create("parser", NULL);
            dispatch_async(queue, ^{
                NSURL *url = [NSURL URLWithString:imageURL];
                NSData *imageData = [NSData dataWithContentsOfURL:url];
                
                dispatch_async(dispatch_get_main_queue(), ^{
                    UIImage *image = [UIImage imageWithData:imageData];
                    if (image) {
                        [cellItem setObject:image forKey:@"thumbImage"];
                        nThumbCount++;
                        
                        if (nThumbCount % 10 == 0) {
                            nLoadedImageCount = nThumbCount;
                        }
                        
                        if (nThumbCount - nLoadedImageCount >= 10) {
                            [self reloadTableView];
                        }
                    }
                });
            });
            
            NSString *currentParkName = [[currentItem objectForKey:@"ParkName"] description];
            
            if ([parkName isEqualToString:currentParkName]) {
                [nearbyArray addObject:cellItem];
            } else {
                currentItem = cellItem;
                [sectionArray addObject:currentItem];
                
                NSArray *itemArray = [NSArray arrayWithArray:nearbyArray];
                [tableviewArray addObject:itemArray];
                [nearbyArray removeAllObjects];
                [nearbyArray addObject:cellItem];
            }
        }
        
        [tableviewArray addObject:nearbyArray];
    }

    [self reloadTableView];
}

- (void)reloadTableView {
    if (tableviewArray.lastObject) {
        [self.aTableView reloadData];
        
        [self.activityView stopAnimating];
    }
}

#pragma mark - UITableViewDataSource

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    if (sectionArray.lastObject) {
        return sectionArray.count;
    }
    else {
        return 0;
    }
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    NSInteger rows = 0;
    
    if ([tableviewArray lastObject]) {
        //NSLog(@"tableView numberOfRowsInSection: %ld", (long)[tableviewArray count]);
        if (tableviewArray.count > section) {
            NSArray *sectionItem = [tableviewArray objectAtIndex:section];
            rows = sectionItem.count;
        } else {
            return 0;
        }
    }
    else {
        return 0;
    }
    
    return rows;
}

- (NSString*)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section
{
    NSString *headerTitle = @"";

    if (sectionArray.lastObject) {
        if (section < sectionArray.count) {
            NSDictionary *sectionItem = [sectionArray objectAtIndex:section];
            headerTitle = [sectionItem objectForKey:@"ParkName"];
        }
    }

    return headerTitle;
}

- (UITableViewCell*)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    ParkListTableViewCell *cell = (ParkListTableViewCell*)[self.aTableView dequeueReusableCellWithIdentifier:@"Park Cell"];

    if (!cell) {
        cell = [[ParkListTableViewCell alloc] initWithStyle:UITableViewCellStyleSubtitle reuseIdentifier:@"Park Cell"];
    }
    
    if (tableviewArray.lastObject) {
        if ([tableviewArray count] > indexPath.row) {
            NSArray *sectionItem = [tableviewArray objectAtIndex:indexPath.section];
            NSDictionary *cellItem = [sectionItem objectAtIndex:indexPath.row];
//            NSString *imageURL = [[cellItem objectForKey:@"Image"] description];
            NSString *parkName = [[cellItem objectForKey:@"ParkName"] description];
            NSString *name = [[cellItem objectForKey:@"Name"] description];
            NSString *introduction = [[cellItem objectForKey:@"Introduction"] description];
//            NSString *openTime = [[cellItem objectForKey:@"OpenTime"] description];
//            NSString *yearBuilt = [[cellItem objectForKey:@"YearBuilt"] description];

            UIImage *image = (UIImage *)[cellItem objectForKey:@"thumbImage"];
            
            if (!image) {
                image = [UIImage imageNamed:@"Round_Landmark_Icon_Park"];
            }
            
            cell.parkImageView.image = image;
            cell.parkNameLabel.text = parkName;
            cell.nameLabel.text = name;
            cell.parkIntroLabel.text = introduction;
        }
    }
    
    return cell;
}


#pragma mark - UITableViewDelegate

- (CGFloat)tableView:(UITableView *)tableView estimatedHeightForRowAtIndexPath:(NSIndexPath *)indexPath
{
    return UITableViewAutomaticDimension;
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath
{
    return UITableViewAutomaticDimension;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    selectedIndexPath = indexPath;
    
    [self performSegueWithIdentifier:@"Show Detail" sender:nil];
    
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
}


 #pragma mark - Navigation
 
 // In a storyboard-based application, you will often want to do a little preparation before navigation
 - (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
 // Get the new view controller using [segue destinationViewController].
 // Pass the selected object to the new view controller.
     
     if ([segue.identifier isEqualToString:@"Show Detail"]) {
         if ([segue.destinationViewController isKindOfClass:[ParkDetailViewController class]]) {
             ParkDetailViewController *pdvc = (ParkDetailViewController*)segue.destinationViewController;
             NSArray *sectionItem = [tableviewArray objectAtIndex:selectedIndexPath.section];
             NSDictionary *cellItem = [sectionItem objectAtIndex:selectedIndexPath.row];
             pdvc.parkItem = cellItem;
             pdvc.nearbyviewArray = sectionItem;
         }
     }
 }




- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}


@end

File: TaiepiParkS/TaiepiParkS/ViewController.swift
//
//  ViewController.swift
//  TaiepiParkS
//
//  Created by Richard on 2018/01/31.
//  Copyright © 2018年 Snowowl. All rights reserved.
//

import UIKit

class ViewController: UIViewController, UITableViewDataSource, UITableViewDelegate, URLSessionDelegate, URLSessionTaskDelegate, URLSessionDataDelegate {

    @IBOutlet weak var aTableView: UITableView! {
        didSet {
            aTableView.dataSource = self
            aTableView.delegate = self
        }
    }
    
    @IBOutlet weak var activityView: UIActivityIndicatorView!
    
    var queryTimer : Timer?
    var timeoutTimer : Timer?
    var timeout : TimeInterval!
    var responseData : NSMutableData?
    
    var sectionArray : NSMutableArray?
    var tableviewArray : NSMutableArray?
    var nearbyviewArray: NSMutableArray?
    
    var selectedIndexPath : IndexPath?
    
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        
        activityView.startAnimating()
        
        self.startQueryTimer()
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        aTableView.rowHeight = UITableViewAutomaticDimension
        aTableView.estimatedRowHeight = 150
    }
    
    // MARK: HTTP Connection
    
    @objc func startConnection() {
        let urlString = Constant.Query_Remote_Url
        let url = URL.init(string: urlString)!
        let request = URLRequest.init(url: url, cachePolicy: URLRequest.CachePolicy.useProtocolCachePolicy, timeoutInterval: Constant.Query_Timeout_Interval)
        let sessionConfiguration = URLSessionConfiguration.default
        let queue = OperationQueue.main
        let session = URLSession(configuration: sessionConfiguration, delegate: self, delegateQueue: queue)
        let task = session.dataTask(with: request as URLRequest!)
        task.resume()
    }
    
    func startQueryTimer() {
        if queryTimer == nil {
            queryTimer = Timer.scheduledTimer(timeInterval: Constant.Query_Timer_Interval, target: self, selector: #selector(self.startConnection), userInfo: nil, repeats: false)
        }
    }
    
    
    // MARK: HTTP NSURLSessionDelegate
    
    func urlSession(_ session: URLSession, dataTask: URLSessionDataTask, didReceive response: URLResponse, completionHandler: @escaping (URLSession.ResponseDisposition) -> Void) {
        let httpResponse = response as! HTTPURLResponse
        let statusCode = httpResponse.statusCode
        
        if statusCode != 200 {
            print("didReceive response error: \(statusCode)")
            print("didReceive response: \(response)")
        } else if statusCode == 200 {
            print("didReceive response successful")
            responseData = NSMutableData.init()
            completionHandler(.allow)
        }
    }
    
    func urlSession(_ session: URLSession, dataTask: URLSessionDataTask, didReceive data: Data) {
        responseData?.append(data)
    }
    
    func urlSession(_ session: URLSession, task: URLSessionTask, didCompleteWithError error: Error?) {
        if error != nil {
            print("didCompleteWithError error: \(error!)")
        }
        else {
            let jsonString = String.init(data: responseData as Data!, encoding: String.Encoding.utf8)
            print("didCompleteWithError data: \(jsonString!)")
            
            self.parseNprocess(with: jsonString!)
        }
    }

    // MARK: Parsing JSON Data
    
    func parseNprocess(with jsonString: String) {
        let jsonData = jsonString.data(using: String.Encoding.utf8)
        let json = try! JSONSerialization.jsonObject(with: jsonData!, options: [JSONSerialization.ReadingOptions.mutableContainers]) as! NSDictionary

        sectionArray = NSMutableArray.init()
        tableviewArray = NSMutableArray.init()
        
        let result = json["result"] as! NSDictionary
        let results = result["results"] as! NSArray
        
        var imageCount = 0
        var thumbCount = 0
        
        if results.lastObject != nil {
            var currentItem = results[0] as! NSDictionary
            sectionArray?.add(currentItem)
            
            let nearbyArray = NSMutableArray.init()
            
            var loadedImageCount = 0
            
            for index in stride(from: 0, to: results.count, by: 1) {
                let item = results[index] as! NSDictionary
                let imageURL = item["Image"] as! String
                let parkName = item["ParkName"] as! String
                let name = item["Name"] as! String
                let introduction = item["Introduction"] as! String
                let openTime = item["OpenTime"] as! String
                let yearBuilt = item["YearBuilt"] as! String
                
                print("\nImage URL: \(imageURL),\nPark Name:\(parkName),\nName: \(name),\nOpen Time: \(openTime),\nYear Build: \(yearBuilt),\nIntroduction: \(introduction)\n")
                
                if imageURL != "" {
                    imageCount += 1
                }
                
                let cellItem = NSMutableDictionary(dictionary: item)
                
                DispatchQueue.global(qos: .userInitiated).async { [weak self] in
                    if let url = URL.init(string: imageURL) {
                        if let imageData = NSData.init(contentsOf: url) as Data? {
                        
                            DispatchQueue.main.async { [weak self] in
                                if let image = UIImage.init(data: imageData) {
                                    cellItem.setValue(image, forKey: "thumbImage")
                                    thumbCount += 1
                                    
                                    if thumbCount % 10 == 0 {
                                        loadedImageCount = thumbCount
                                    }
                                    
                                    if thumbCount - loadedImageCount >= 10 {
                                        self!.reloadTableView()
                                    }
                                }
                            }
                        }
                    }
                }
                
                let currentParkName = currentItem["ParkName"] as! String
                
                if parkName == currentParkName {
                    nearbyArray.add(cellItem)
                } else {
                    currentItem = cellItem
                    sectionArray?.add(currentItem)
                    
                    let itemArray = NSArray.init(array: nearbyArray)
                    tableviewArray?.add(itemArray)
                    nearbyArray.removeAllObjects()
                    nearbyArray.add(cellItem)
                }
            }
            
            tableviewArray?.add(nearbyArray)
        }
        
        reloadTableView()
    }
    
    func reloadTableView() {
        DispatchQueue.main.async { [weak self] in
            if self!.tableviewArray?.lastObject != nil {
                self!.aTableView.reloadData()
                
                self!.activityView.stopAnimating()
            }
        }
    }
    
    
    // MARK: - UITableViewDataSource
    
    func numberOfSections(in tableView: UITableView) -> Int {
        if sectionArray?.lastObject != nil {
            print("sections: \(sectionArray!.count)")
            return sectionArray!.count
        } else {
            return 0
        }
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        var rows = 0
        if tableviewArray!.lastObject != nil {
            if tableviewArray!.count > section {
                if let sectionItem = tableviewArray![section] as? NSArray {
                    rows = sectionItem.count
                    print("rows: \(rows) in section: \(section)")
                }
            } else {
                return 0
            }
        }
        else {
            return 0
        }
        
        return rows
    }
    
    func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        var headerTitle = ""

        if sectionArray!.lastObject != nil {
            if section < sectionArray!.count {
                if let sectionItem = sectionArray![section] as? NSDictionary {
                    headerTitle = sectionItem["ParkName"] as! String
                    print("title: \(headerTitle) of section: \(section)")
                }
            }
        }
        return headerTitle
    }

    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        var cell = aTableView.dequeueReusableCell(withIdentifier: "Park Cell") as? ParkListTableViewCell

        if cell == nil {
            cell = UITableViewCell.init(style: UITableViewCellStyle.subtitle, reuseIdentifier: "Park Cell") as? ParkListTableViewCell
        }
        
        if tableviewArray!.lastObject != nil {
            if tableviewArray!.count > indexPath.row {
                if let sectionItem = tableviewArray![indexPath.section] as? NSArray {
                    if let cellItem = sectionItem[indexPath.row] as? NSDictionary {
                        let parkName = cellItem["ParkName"] as! String
                        let name = cellItem["Name"] as! String
                        let introduction = cellItem["Introduction"] as! String
                        
                        var image = cellItem["thumbImage"] as? UIImage
                        
                        if image == nil {
                            image = UIImage.init(named: "Round_Landmark_Icon_Park")
                        }
                        
                        cell?.parkImageView.image = image
                        cell?.parkNameLabel.text = parkName
                        cell?.nameLabel.text = name
                        cell?.parkIntroLabel.text = introduction
                    }
                }
            }
        }

        return cell!
    }
    
    // MARK: - UITableViewDelegate
    
    func tableView(_ tableView: UITableView, estimatedHeightForRowAt indexPath: IndexPath) -> CGFloat {
        return UITableViewAutomaticDimension
    }
    
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return UITableViewAutomaticDimension
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        selectedIndexPath = indexPath
        
        performSegue(withIdentifier: "Show Detail", sender: nil)

        tableView.deselectRow(at: indexPath, animated: true)
    }


    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

     // MARK: - Navigation
     
     // In a storyboard-based application, you will often want to do a little preparation before navigation
     override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
     // Get the new view controller using segue.destinationViewController.
     // Pass the selected object to the new view controller.
        
        if segue.identifier == "Show Detail" {
            if segue.destination is ParkDetailViewController {
                if let pdvc = segue.destination as? ParkDetailViewController {
                    let sectionItem = tableviewArray![selectedIndexPath!.section] as! NSArray
                    let cellItem = sectionItem[selectedIndexPath!.row] as! NSDictionary
                    pdvc.parkItem = cellItem
                    pdvc.nearbyviewArray = sectionItem
                }
            }
        }
        
     }
    

}

struct Constant {
    static let Query_Remote_Url = "http://data.taipei/opendata/datalist/apiAccess?scope=resourceAquire&rid=bf073841-c734-49bf-a97f-3757a6013812"
    static let Query_Timer_Interval = 3.0
    static let Query_Timeout_Interval = 15.0
}

